system/ioctl.go

```diff
@@ -240,65 +195,39 @@ func getUbiDeviceSize(file *os.File) (uint64, error) {
 	return reservedSectors * sectorSize, nil
 }
 
-func GetBlockDeviceSectorSize(file *os.File) (int, error) {
-	var sectorSize int
-	var err error
-
-	_, _, errno := sys.RawSyscall(
-		uintptr(unix.SYS_IOCTL),
-		file.Fd(),
-		uintptr(unix.BLKSSZGET),
-		uintptr(unsafe.Pointer(&sectorSize)))
+// GetDeviceIDFromPath retrieves the device id for the block device pointed to by
+// the inode at path.
+func GetDeviceIDFromPath(path string) ([2]uint32, error) {
+	var stat stat
 
-	if errno != 0 {
-		// ENOTTY: Inappropriate I/O control operation - in this context
-		// it means that the file descriptor is not a block-device
-		if errno == unix.ENOTTY {
-			// Check if it is an UBI block device
-			sectorSize, err = getUbiDeviceSectorSize(file)
-			if err != nil {
-				return 0, err
-			}
-		} else {
-			return 0, errors.New(errno.Error())
-		}
+	if err := sys.Stat(path, &stat); err != nil {
+		return [2]uint32{^uint32(0), ^uint32(0)}, errors.Wrapf(err,
+			"stat %s", path)
 	}
-	return sectorSize, nil
-}
 
-func GetBlockDeviceSize(file *os.File) (uint64, error) {
-	var devSize uint64
-	var err error
-	_, _, errno := sys.RawSyscall(
-		uintptr(unix.SYS_IOCTL), uintptr(file.Fd()),
-		uintptr(unsafe.Pointer(uintptr(unix.BLKGETSIZE64))),
-		uintptr(unsafe.Pointer(&devSize)))
+	devType := stat.Mode & unix.S_IFMT
 
-	if errno != 0 {
-		// ENOTTY: Inappropriate I/O control operation - in this context
-		// it means that the file descriptor is not a block-device
-		if errno == unix.ENOTTY {
-			// Check if it is an UBI block device
-			devSize, err = getUbiDeviceSize(file)
-			if err != nil {
-				return 0, err
-			}
-		} else {
-			return 0, errors.New(errno.Error())
-		}
-	}
-	return devSize, nil
-}
+	switch devType {
+	// If path refers to a special file (e.g. device file under /dev), then
+	// st_dev refers to the device number of the mounted devfs. The device
+	// number for a special file is under the st_rdev property, ref stat(2).
+	case unix.S_IFBLK,
+		unix.S_IFCHR,
+		unix.S_IFIFO,
+		unix.S_IFSOCK:
+		return [2]uint32{
+			unix.Major(uint64(stat.Rdev)), unix.Minor(uint64(stat.Rdev))}, nil
 
-// FreezeFS freezes the filesystem for which the inode that fd points to belongs
-// to, maintaining read-consistency. All write operations to the filesystem will
-// be blocked until ThawFS is called.
-func FreezeFS(fd int) error {
-	err := sys.IoctlSetInt(fd, IOCTL_FIFREEZE_MAGIC, 0)
-	if err != nil {
-		return errors.Wrap(err, "error freezing fs from writing")
+	// If path refers to a regular file, then st_dev gives the device number
+	// of the underlying block device.
+	case unix.S_IFDIR,
+		unix.S_IFREG,
+		unix.S_IFLNK:
+		return [2]uint32{
+			unix.Major(uint64(stat.Dev)), unix.Minor(uint64(stat.Dev))}, nil
 	}
-	return nil
+	return [2]uint32{^uint32(0), ^uint32(0)},
+		fmt.Errorf("invalid stat(2) st_mode %04X", devType)
 }
```

Comment:
This file is really hard to review. As we have moved the functions around the diff shows many functions are removed and added. I looked and some of the functions to see that there were no code differences, so I assume that none of them had code differences :)